src/v/redpanda/admin/security.cc

```diff
@@ -723,3 +722,32 @@ admin_server::update_role_members_handler(
     co_await throw_on_error(*req, err, model::controller_ntp);
     co_return ss::json::json_return_type(j_res);
 }
+
+ss::future<ss::json::json_return_type>
+admin_server::list_roles_handler(std::unique_ptr<ss::http::request> req) {
+    auto filter = req->get_query_param("filter");
+    auto user = req->get_query_param("user");
+
+    auto pred = [&filter, &user](const auto& role_entry) {
+        auto name_prefix_matches = filter.empty()
+                                   || security::role_store::name_prefix_filter(
+                                     role_entry, filter);
+        auto role_has_user = user.empty()
+                             || security::role_store::has_member(
+                               role_entry,
+                               security::role_member{
+                                 security::role_member_type::user, user});
+        return name_prefix_matches && role_has_user;
+    };
+
+    auto roles = _controller->get_role_store().local().range(pred);
+
+    ss::httpd::security_json::roles_list j_res{};
+    for (const auto& role_name : roles) {
+        ss::httpd::security_json::role_description j_desc;
+        j_desc.name = ss::sstring{role_name};
+        j_res.roles.push(j_desc);
+    }
+
+    return ss::make_ready_future<ss::json::json_return_type>(j_res);
```

Comment:
> This generated code isn't very nice to work with; it could do with some constructors, and move semantics. and a push_back, so that you can use it with std::transform and std::back_inserter (although we could easily write a json::pusher that does calls push on the list), and then not copy the whole thing into the response.

Do you think it's worth a backlog ticket to make this change to seastar?

> And if this is expected to get large, hopefully the array is streamed?

It does look like it is streamed: https://github.com/scylladb/seastar/blob/cd8a9133d2c02f63dbd578d882cf7333a427e194/include/seastar/json/formatter.hh#L103-L118

Reply:
>generated code isn't very nice to work with

fwiw, I've been working on changing this in spare time over the last few days. lots of room for improvement.